### PR TITLE
fix(core): hide `dependencyRoots` from public API

### LINKED ISSUE #2348: Hide Construct.dependencyRoots
We wanted to ensure that the base Construct public API is clean (besides `node`) and leaves room for the API of the actual construct. Somehow `dependencyRoots` snuck in.

Is there a way to get rid of this? Why even use an override here? Can't we do something like:

```ts
construct.node.addDependencyRoot(x);
```

### PR DESCRIPTION
**Pull request by rix0rrr**
`IDependable` is now a marker interface, and signals that there is
a `DependableTrait` implementation which can be retrieved which contains
the actual implementation of `dependencyRoots` for this instance.

Fixes #2348.

----

### Pull Request Checklist

* [ ] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ ] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [ ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/cdk/lib/construct.ts [hunk 1]
```diff
@@ -596,6 +588,11 @@ export class Construct implements IConstruct {
    */
   constructor(scope: Construct, id: string) {
     this.node = new ConstructNode(this, scope, id);
+
+    const self = this;
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/construct.ts]: put a wee comment explaining what this does

File: packages/@aws-cdk/cdk/lib/dependency.ts [hunk 2]
```diff
@@ -27,17 +24,61 @@ export interface IDependable {
...
+
+/**
+ * Trait for IDependable
+ *
+ * Traits are interfaces that are privately implemented by objects. Instead of
+ * showing up in the public interface of a class, they need to be queried
+ * explicitly. This is used to implement certain framework features that are
+ * not intended to be used by Construct consumers, and so should be hidden
+ * from accidental use.
+ */
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/dependency.ts]: Example on how this is used

File: packages/@aws-cdk/cdk/lib/dependency.ts [hunk 3]
```diff
@@ -27,17 +24,61 @@ export interface IDependable {
 export class ConcreteDependable implements IDependable {
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/dependency.ts]: Where is this used?

Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/dependency.ts]: https://github.com/awslabs/aws-cdk/search?q=concretedependable&unscoped_q=concretedependable

Comment (eladb) [reply packages/@aws-cdk/cdk/lib/dependency.ts]: Oh nice. for some reason I thought it is just introduced now.

**Pull request merged**